### DISCUSSION TITLE
Add hash fields for files (md5/sha256/...)

### DISCUSSION THREAD

**Discussion by andrenarchy** [directus/directus#5162] [Feature Requests]
It would be great if Directus would calculate hash(es) of uploaded files automatically and make the hash(es) available through the file's fields. My use case is that I sync some files to another system frequently and having hashes would allow me to skip files that are already up to date. I suggest adding md5 and sha256 hashes which are very common.

The hashing could be an opt-in feature but given that the hashing usually is not the bottleneck but rather i/o (especially with cloud storage) I'd even say that it could just be an always-on feature.

What do you think?

**Comment by benhaynes** [comment]
We had this in v8, and maybe even in v9. I think we even had it in v9 for a while, but it was removed... maybe for performance reasons? @rijkvanzanten ?

**Reply by andrenarchy** [reply]
Oh that's interesting! If performance was the issue then we could make it an opt-in feature maybe?

**Reply by benhaynes** [reply]
I think we had it only process for files under a certain size, but I defer to Rijk on the actual reason.

**Reply by rijkvanzanten** [reply]
Directus itself didn't use it for anything, and we didn't get any signals that anybody was actually using it for anything at the time, so it was dropped

I'm down to look into making it a thing in core again (assuming it meets 80/20), but this could also be a user-land thing through a hook 🤔

**Reply by andrenarchy** [reply]
How would a hook look like for this kind of thing? Is there an example somewhere for a hook that works on files?

**Reply by rijkvanzanten** [reply]
I think the (to be added) `files.upload` hook would be perfect for this 🤔 If we can pass the file stream to that hook, you should be able to extract and save the checksum

**Reply by andrenarchy** [reply]
Ah I see, so currently it's not yet possible?

By the way, one more argument for adding hashes to core would be that it'd allow Directus to spot duplicates. For example, it could display a warning in the future if you try to upload the same file twice (which already happened to me a couple of times...).

**Reply by benhaynes** [reply]
Identifying duplicates was my main reason for wanting this too.

**Reply by ajira86** [reply]
This is a good reason, I know that [DatoCMS use it to detect duplicates and warn user](https://www.datocms.com/product-updates/duplicate-asset-detection).

**Comment by rgieseke** [comment]
Where would this be stored, in a `metadata` field? If one adds such a hook or calculates the hash locally?
Is `metadata` safe to write to or is it just for extracted metadata?

**Reply by benhaynes** [reply]
The metadata field might be overridden when the platform scrapes the asset for IPTC/EXIF/etc. I would recommend making a custom checksum field.

**Reply by rgieseke** [reply]
Thanks! Also just saw this in the docs, which is quite clear (previously i had only looked at the API response)
https://docs.directus.io/concepts/files/

>The sidebar's info component also includes the following readonly details:
> 
>     Type — The MIME type of the file, displayed in the App as a formatted media type
>     Dimensions — (Images Only) The width and height of the image in pixels
>     Size — The file-size the asset takes up within the storage adapter
>     Created — The timestamp of when the file was uploaded to the project
>     Owner — The Directus user that uploaded the file to the project
>     Folder — The current parent folder that contains the file
>     Metadata — Metadata JSON dump of the file's EXIF, IPTC, and ICC information

So it's better to store that in a separate collection then?

**Reply by benhaynes** [reply]
Well, if you're building this custom, you can add a custom field to `directus_files` and that should work. But this is also a feature request to add this into Directus Core... in which case we would add a similar field ourselves.

**Reply by rgieseke** [reply]
Great, just would need to check and write a custom migration if there were a clash. Thanks again for the quick replies!

**Comment by lgvld** [comment]
Still an interesting idea.

The current behavior of Directus allows an user to upload the same file any number of times (with the exact same name), without any warning. Why would this be the default behavior?

One should create a `file.upload` hook in order to prevent duplicates? i.e. fill a checksum field and compare with all existing checksums on upload

What I miss when working with files in Directus is the very basic behavior of a S3 bucket, where the name of the file is in a sense its primary key. Does Directus provide a S3 endpoint mapping the underlying S3 bucket (or a "subdirectory" in the bucket) used for uploads?

Thanks.

;-)

**Reply by rijkvanzanten** [reply]
> Why would this be the default behavior?

Because there's no guarantees on intended behavior from the end user. If I create a new "report" on my computer, and save it as the default `Untitled.pages`, I wouldn't expect Directus to start erroring out when trying to upload my new report for example. That being said, I agree it would be nice if there was an out of the box way for Directus to handle duplicates more gracefully 👍🏻

**Reply by lgvld** [reply]
You’re right, thank you. This behavior is actually the most flexible/general. But maybe too much for my basic usecase. ;-)

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by ralphwest1** [comment]
Does anyone know if there is community plugin that will add the md5 hash to uploads that can be used until this feature is added to core?

**Comment by jacob-israel-turner** [comment]
Hi Directus team! Is there any plan to move forward on this feature? Preventing duplicates in the file system is a really high-value feature for my project, and hashes seem like the most straightforward way to accomplish this.

Thanks for your hard work on this project!

**Reply by rijkvanzanten** [reply]
Heya 👋🏻 I'd like to have it some day, but simultaneously have my plate pretty full with other things, so can't make any promises on timelines when this will make it's way to the top of the todo list..

Sounds like this feature request would be a starting point, but not a full solution yet for your use case. Implementing support for file hashes is relatively straightforward, but where this gets trickier is how to deal with duplicates and make sure that is handled properly. 

What do you imagine the user flow to look like when a duplicate file is being uploaded? Would you expect an option to be presented to let the user choose between "Nevermind, use existing" and "upload anyways"? What happens if you want to make file modifications (like crop/rotate), but suddenly the file is shared across multiple places regardless? 🤔

**Reply by jacob-israel-turner** [reply]
> What do you imagine the user flow to look like when a duplicate file is being uploaded?

I imagined being able to set a unique constraint on the folder level. If a unique constraint is set, then when the user uploads a duplicate file, they are simply given an error, possibly with a link to the already-existing asset.

> Would you expect an option to be presented to let the user choose between "Nevermind, use existing" and "upload anyways"?

I don't have the need for an "upload anyways" flow. Perhaps, if the unique constraint is not set, you could throw up an option like this.

> What happens if you want to make file modifications (like crop/rotate), but suddenly the file is shared across multiple places regardless?

I'm not sure I understand this question - does Directus allow you to crop/rotate an image in the app? If so, surely you'd have to re-generate the hash, right? I imagine it would be considered a new file if it's been modified, but I suppose that depends on how the transformation has been applied/saved. I am not familiar with the specifics.

I'm not using images, but documents, so my use-cases might be different than what you're imagining. I believe that simply generating the hashes, and ideally allowing for a unique constraint in the file library, would be a solution. I am mostly using the SDK to upload files, not the app - so as long as the API returns a reasonable error when uploading a duplicate file, I can handle that properly.

**Comment by burka** [comment]
It would be great to have this to add robust ETag support and provide good caching headers.

**Comment by paescuj** [comment]
To add some context, we did look into this a few months ago, but decided against it at that time for the following reason:
https://github.com/directus/directus/issues/22606#issuecomment-2228785180

**Reply by jacob-israel-turner** [reply]
> The main goal of this implementation is to have the flag to indicate whether the file was changed, to solve that we don't need full checksumming. Juice ain't worth the squeeze 🙂

The use-case here is different, and checksumming seems to be the only solution.

Besides not being worth the squeeze, was checksumming infeasible altogether? I understand performance concerns, but for some of us that tradeoff is potentially worthwhile.
